spring-integration-core/src/main/java/org/springframework/integration/channel/management/AggregatingMessageChannelMetrics.java

```diff
@@ -0,0 +1,98 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package org.springframework.integration.channel.management;
+
+import org.springframework.integration.support.management.ExponentialMovingAverage;
+import org.springframework.integration.support.management.ExponentialMovingAverageRate;
+import org.springframework.integration.support.management.ExponentialMovingAverageRatio;
+import org.springframework.integration.support.management.MetricsContext;
+
+/**
+ * Implementation that provides the total response times over a sample, to avoid fetching the system time
+ * on every message.
+ *
+ * @author Gary Russell
+ * @since 4.2
+ */
+public class AggregatingMessageChannelMetrics extends DefaultMessageChannelMetrics {
+
+	private static final int DEFAULT_SAMPLE_SIZE = 1000;
+
+	private final int sampleSize;
+
+	public AggregatingMessageChannelMetrics() {
+		this(null, DEFAULT_SAMPLE_SIZE);
+	}
+
+	/**
+	 * Construct an instance with default metrics with {@code window=10, period=1 second,
+	 * lapsePeriod=1 minute}.
+	 * @param name the name.
+	 * @param sampleSize the sample size
+	 */
+	public AggregatingMessageChannelMetrics(String name, int sampleSize) {
+		super(name);
+		this.sampleSize = sampleSize;
+	}
+
+	/**
+	 * Construct an instance with the supplied metrics. For proper representation of metrics, the
+	 * supplied sendDuration must have a {@code factor=1000000.} and the the other arguments
+	 * must be created with the {@code millis} constructor argument set to true.
+	 * @param name the name.
+	 * @param sendDuration an {@link ExponentialMovingAverage} for calculating the send duration.
+	 * @param sendErrorRate an {@link ExponentialMovingAverageRate} for calculating the send error rate.
+	 * @param sendSuccessRatio an {@link ExponentialMovingAverageRatio} for calculating the success ratio.
+	 * @param sendRate an {@link ExponentialMovingAverageRate} for calculating the send rate.
+	 * @param sampleSize the sample size
+	 * @since 4.2
+	 */
+	public AggregatingMessageChannelMetrics(String name, ExponentialMovingAverage sendDuration,
+			ExponentialMovingAverageRate sendErrorRate, ExponentialMovingAverageRatio sendSuccessRatio,
+			ExponentialMovingAverageRate sendRate, int sampleSize) {
+		super(name, sendDuration, sendErrorRate, sendSuccessRatio, sendRate);
+		this.sampleSize = sampleSize;
+	}
+
+	@Override
+	public synchronized MetricsContext beforeSend() {
+		long start = 0;
+		if (this.sendCount.getAndIncrement() % this.sampleSize == 0 && isFullStatsEnabled()) {
```

Comment:
maybe reverse?
```
(isFullStatsEnabled() && this.sendCount.getAndIncrement() % this.sampleSize == 0)
```
Because we always `getAndIncrement()` independently of full stat enablement.
Have I missed anything?